core/types/src/attestation_key.rs

```diff
@@ -0,0 +1,272 @@
+    new_type_accessors_impls,
+    report::{ExtendedProductId, FamilyId},
+    svn::ConfigId,
+    FfiError,
+};
+use mc_sgx_core_sys_types::{sgx_att_key_id_ext_t, sgx_ql_att_key_id_t, sgx_quote_sign_type_t};
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub enum MrSignerKeyHash {
+    Sha256([u8; 32]),
+    Sha384([u8; 48]),
+}
+
+#[derive(Debug, Default, Clone, Hash, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct AlgorithmId(u32);
+
+new_type_accessors_impls! {
+    AlgorithmId, u32;
+}
+
+#[derive(Debug, Default, Clone, Hash, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct Version(u16);
+
+new_type_accessors_impls! {
+    Version, u16;
+}
+
+#[derive(Debug, Default, Clone, Hash, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct Id(u16);
+
+new_type_accessors_impls! {
+    Id, u16;
+}
+
+/// The type of quote signature.  Only valid for EPID quotes.
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+#[non_exhaustive]
+#[repr(u16)]
+enum QuoteSignature {
+    UnLinkable,
+    Linkable,
+}
+
+impl From<sgx_quote_sign_type_t> for QuoteSignature {
+    fn from(sign_type: sgx_quote_sign_type_t) -> QuoteSignature {
+        match sign_type {
+            // Per the header `sgx_quote.h` and
+            // https://download.01.org/intel-sgx/sgx-linux/2.17.1/docs/Intel_SGX_Developer_Reference_Linux_2.17.1_Open_Source.pdf
+            // the `sgx_att_key_id_ext_t::att_key_type` is only valid for EPID
+            // quotes it will be 0 otherwise, which also happens to map to
+            // SGX_UNLINKABLE_SIGNATURE.
+            sgx_quote_sign_type_t::SGX_UNLINKABLE_SIGNATURE => UnLinkable,
+            sgx_quote_sign_type_t::SGX_LINKABLE_SIGNATURE => Linkable,
+            v => panic!("Got unknown 'sgx_quote_sign_type_t': {:?}", v),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct QuoteLibAttestationKeyId(sgx_ql_att_key_id_t);
+
+impl QuoteLibAttestationKeyId {
+    /// The ID
+    pub fn id(&self) -> Id {
+        self.0.id.into()
+    }
+
+    /// The key version
+    pub fn version(&self) -> Version {
+        self.0.version.into()
+    }
+
+    /// The hash of the MRSIGNER key.
+    ///
+    /// # Errors
+    /// When the length of the hash is not a valid length for a hash,
+    /// [`sgx_ql_att_key_id_t::mrsigner_length`].
+    pub fn mr_signer_key_hash(&self) -> Result<MrSignerKeyHash, FfiError> {
+        match self.0.mrsigner_length {
+            32 => {
+                let mut array: [u8; 32] = [0u8; 32];
+                array.copy_from_slice(&self.0.mrsigner[..32]);
+                Ok(MrSignerKeyHash::Sha256(array))
+            }
+            48 => {
+                let mut array: [u8; 48] = [0u8; 48];
+                array.copy_from_slice(&self.0.mrsigner[..48]);
+                Ok(MrSignerKeyHash::Sha384(array))
+            }
+            _ => Err(FfiError::InvalidInputLength),
+        }
+    }
+
+    /// The product ID
+    pub fn product_id(&self) -> ExtendedProductId {
```

Comment:
Intentionally skipped the `prod_id` member since it said it was for legacy.  I think IFF we want it then we add logic here to map into the `ExtendedProductId`.  We could use an enum but that seems like more work for the end user than hopefully necessary.

Also I reused the `ExtendedProductId` from `report.rs`.  They are the same length and I'm pretty sure they're meant to be the same, but for some reason this one here in quote used raw sizes

Reply:
We're currently using product ID in all our enclaves, so we will need to expose them both.

Reply:
Doing a quick look through the  MobileCoin repo I'm not seeing this product id used.
> prod_id
Legacy Product ID of the QE.
extended_prod_id
Extended Product ID of the QE. All 0's for legacy format enclaves.

I'm not seeing `sgx_ql_att_key_id_t` used anywhere in the MobileCoin repo.